SQL Store.adoc

```diff
@@ -0,0 +1,251 @@
+         <queries key-columns="isbn">
+            <select-all>SELECT isbn, title FROM books</select-all>
+            <select>SELECT isbn, title FROM books WHERE isbn = :key</select>
+            <delete-all>DELETE FROM books</delete-all>
+            <delete>DELETE FROM books WHERE isbn = :key</delete>
+            <upsert>INSERT INTO books (isbn, title) VALUES (:key, :value) ON CONFLICT (isbn) DO UPDATE SET title = :value</upsert>
+         </queries>
+         <!-- the schema element is optional. Defaults will be used otherwise -->
+         <schema name="books.proto" key-name="books_id.proto" package="sql_store"/>
+      </query-jdbc-store>
+   </persistence>
+</local-cache>
+----
+
+==== Query parameters
+
+Query parameters are named using the following conventions:
+
+* primitive keys or values use the `:key` or `:value` names
+* composite keys or values use the `:key` or `:value` names followed by a `.` and the name of the field, e.g. `:value.description`
+
+== Storage
+
+=== ProtoBuf schema
+
+==== User-supplied schema
+
+It should be possible for the user to specify her own schema name, which should be present in the schema registry. 
+The names of the fields in the schema **MUST** match the column names of the table.
+
+==== Generated schema
+
+===== Naming
+
+The default naming of the schema will use the following convention: `*cache-name*-sql-store.proto`.
+
+=== Type mapping
+
+==== Primitive
+
+Primitive type mapping is used when the table/query only returns two columns. The following conventions are used:
+
+* A string will be used for column types which can be represented as such (char, varchar, text, numbers, boolean)
+* a byte array will be used for binary types (blob, bytea)
+
+
+==== ProtoBuf
+
+The following table shows the default mappings between common SQL types and their ProtoBuf equivalents:
+
+[cols="1,1"]
+|===
+| *SQL*
+| *ProtoBuf*
+
+| `int4`
+| `int32`
+
+| `int8`
+| `int64`
+
+| `float4`
+| `float`
+
+| `float8`
+| `double`
+
+| `numeric`
+| `double`
+
+| `bool`
+| `bool`
+
+| `char`
+| `string`
+
+| `varchar`
+| `string`
+
+| `text`,`tinytext`,`mediumtext`,`longtext`
+| `string`
+
+| `bytea`,`tinyblob`,`blob`,`mediumblob`,`longblob`
+| `bytes`
+
+|===
+
+== Examples
+
+=== Two columns: Primitive key and value
+
+[source,sql]
+----
+CREATE TABLE books (
+    isbn NUMBER(13),
+    title varchar(120)
+    PRIMARY KEY(isbn)
+);
+----
+
+the default behaviour will cause the `isbn` column to be used as key and the `title` column to be used as value, both as strings. No ProtoBuf schemas will be generated. The following queries will be used (PostgreSQL):
+
+* SELECT: `SELECT isbn, title FROM books WHERE isbn = ?`
+* SELECT-ALL: `SELECT isbn, title FROM books`
+* DELETE: `DELETE FROM books WHERE isbn = ?`
+* DELETE-ALL: `TRUNCATE books`
+* UPSERT: `INSERT INTO books (isbn, title) VALUES (?, ?) ON CONFLICT (isbn) DO UPDATE SET title = ?`
```

Comment:
I suggest using the `:key` and `:value` parameter names here in the other examples of generated SQL, to match what the user would have to write in the configuration.

Reply:
Done